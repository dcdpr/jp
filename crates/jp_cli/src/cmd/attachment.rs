use jp_attachment_bear_note as _;
use jp_attachment_cmd_output as _;
use jp_attachment_file_content as _;
use jp_conversation::Context;
use tracing::{debug, trace};
use url::Url;

use super::Output;
use crate::{
    ctx::Ctx,
    error::{Error, Result},
};

pub(super) mod add;
mod ls;
mod rm;

#[derive(Debug, clap::Args)]
pub struct Args {
    #[command(subcommand)]
    command: Commands,
}

impl Args {
    pub async fn run(self, ctx: &mut Ctx) -> Output {
        match self.command {
            Commands::Add(args) => args.run(ctx).await,
            Commands::Remove(args) => args.run(ctx).await,
            Commands::List(args) => args.run(ctx).await,
        }
    }
}

#[derive(Debug, clap::Subcommand)]
enum Commands {
    /// Add attachment to context.
    #[command(name = "add", alias = "a")]
    Add(add::Args),

    /// Remove attachment from context
    #[command(name = "rm", alias = "r")]
    Remove(rm::Args),

    /// List attachments in context.
    #[command(name = "ls", alias = "l")]
    List(ls::Args),
}

pub async fn register_attachment(uri: &Url, ctx: &mut Context) -> Result<()> {
    trace!(uri = uri.as_str(), "Registering attachment.");

    let scheme = uri.scheme();
    let attachment = if let Some(attachment) = ctx.attachment_handlers.get_mut(scheme) {
        attachment
    } else {
        let Some(handler) = jp_attachment::find_handler_by_scheme(scheme) else {
            return Err(Error::NotFound("Attachment handler", scheme.to_string()));
        };

        ctx.attachment_handlers
            .entry(scheme.to_string())
            .or_insert(handler)
    };

    debug!(%uri, "Registered URI as attachment.");
    attachment
        .add(uri)
        .await
        .map_err(|e| Error::Attachment(e.to_string()))
}

pub async fn unregister_attachment(uri: &str, ctx: &mut Context) -> Result<()> {
    let uri = if let Ok(uri) = Url::parse(uri) {
        uri
    } else {
        // Special case for file attachments
        trace!("URI is not a valid URL, treating as file path.");
        Url::parse(&format!("file:{uri}"))?
    };

    let id = uri.scheme();

    if let Some(attachment) = ctx.attachment_handlers.get_mut(id) {
        attachment
            .remove(&uri)
            .await
            .map_err(|e| Error::Attachment(e.to_string()))?;
    }

    Ok(())
}
