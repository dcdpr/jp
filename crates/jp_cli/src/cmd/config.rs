use std::ffi::OsStr;

use jp_config::{
    fs::{ConfigFile, ConfigLoader, ConfigLoaderError},
    user_global_config_path,
};

use super::Output;
use crate::ctx::Ctx;

mod fmt;
mod set;
mod show;

#[derive(Debug, clap::Args)]
pub(crate) struct Config {
    #[command(subcommand)]
    command: Commands,
}

impl Config {
    pub(crate) fn run(self, ctx: &mut Ctx) -> Output {
        match self.command {
            Commands::Show(args) => args.run(ctx),
            Commands::Set(args) => args.run(ctx),
            Commands::Fmt(args) => args.run(ctx),
        }
    }
}

#[derive(Debug, clap::Subcommand)]
enum Commands {
    /// Show the current configuration.
    #[command(name = "show")]
    Show(show::Show),

    /// Set a configuration option.
    #[command(name = "set")]
    Set(set::Set),

    /// Format a configuration file.
    #[command(name = "fmt")]
    Fmt(fmt::Fmt),
}

#[derive(Debug, Clone, Copy, Default, PartialEq, clap::Args)]
#[group(required = false, multiple = false)]
struct Target {
    /// The workspace's user-specific configuration target.
    #[arg(long)]
    pub(crate) user_workspace: bool,

    /// The global user-specific configuration target.
    #[arg(long)]
    pub(crate) user_global: bool,

    /// The current-working-directory configuration target, recursively upwards.
    #[arg(long)]
    pub(crate) cwd: bool,
}

impl Target {
    /// Get the the configuration file, if one exists.
    pub(crate) fn config_file(self, ctx: &Ctx) -> Result<Option<ConfigFile>, ConfigLoaderError> {
        let mut loader = ConfigLoader {
            file_stem: "config".into(),
            ..Default::default()
        };

        if self.user_workspace {
            ctx.workspace
                .user_storage_path()
                .map(|p| loader.load(p))
                .transpose()
        } else if self.user_global {
            user_global_config_path(std::env::home_dir().as_deref())
                .map(|mut p| {
                    if p.is_file()
                        && let Some(stem) = p.file_name().and_then(OsStr::to_str)
                        && let Some(path) = p.parent()
                    {
                        loader.file_stem = stem.to_owned().into();
                        p = path.to_path_buf();
                    }

                    loader.load(p)
                })
                .transpose()
        } else if self.cwd {
            loader.file_stem = ".jp".into();
            loader.recurse_up = true;
            loader.recurse_stop_at = Some(ctx.workspace.root.clone());

            loader.load(std::env::current_dir()?).map(Some)
        } else {
            ctx.workspace
                .storage_path()
                .map(|p| loader.load(p))
                .transpose()
        }
    }
}

#[derive(Debug, Default, clap::Args)]
#[group(required = false, multiple = false)]
struct TargetWithConversation {
    #[command(flatten)]
    target: Target,

    /// Appply the configuration to the active or specified conversation.
    #[arg(long, value_name = "CONVERSATION_ID")]
    conversation: Option<Option<String>>,
}
