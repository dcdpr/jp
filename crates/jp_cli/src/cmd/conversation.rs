use super::Output;
use crate::ctx::Ctx;

mod edit;
mod ls;
mod rm;
mod show;
mod use_;

#[derive(Debug, clap::Args)]
pub struct Args {
    #[command(subcommand)]
    command: Commands,
}

impl Args {
    pub fn run(self, ctx: &mut Ctx) -> Output {
        match self.command {
            Commands::Show(args) => args.run(ctx),
            Commands::Remove(args) => args.run(ctx),
            Commands::List(args) => args.run(ctx),
            Commands::Use(args) => args.run(ctx),
            Commands::Edit(args) => args.run(ctx),
        }
    }
}

#[derive(Debug, clap::Subcommand)]
enum Commands {
    /// Remove conversations.
    #[command(name = "rm")]
    Remove(rm::Args),

    /// List conversations.
    #[command(name = "ls")]
    List(ls::Args),

    /// Show conversation details.
    #[command(name = "show")]
    Show(show::Args),

    /// Set the active conversation.
    #[command(name = "use")]
    Use(use_::Args),

    /// Edit conversation details.
    #[command(name = "edit")]
    Edit(edit::Args),
}
