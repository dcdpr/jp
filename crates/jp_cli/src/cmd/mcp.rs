use super::Output;
use crate::ctx::Ctx;

mod attach;
mod detach;
mod edit;
mod list;
mod setup;

#[derive(Debug, clap::Args)]
pub struct Args {
    #[command(subcommand)]
    command: Commands,
}

impl Args {
    pub fn run(self, ctx: &mut Ctx) -> Output {
        match self.command {
            Commands::Setup(args) => args.run(ctx),
            Commands::Attach(args) => args.run(ctx),
            Commands::Detach(args) => args.run(ctx),
            Commands::List(args) => args.run(ctx),
            Commands::Edit(args) => args.run(ctx),
        }
    }
}

#[derive(Debug, clap::Subcommand)]
enum Commands {
    /// Add an MCP server configuration
    #[command(name = "setup")]
    Setup(setup::Args),

    /// Edit (or create) an MCP server configuration in your editor
    #[command(name = "edit")]
    Edit(edit::Args),

    /// Attach an MCP server to the current conversation
    #[command(name = "attach", alias = "a")]
    Attach(attach::Args),

    /// Detach an MCP server from the current conversation
    #[command(name = "detach", alias = "d")]
    Detach(detach::Args),

    /// List all MCP servers
    #[command(name = "ls")]
    List(list::Args),
}
