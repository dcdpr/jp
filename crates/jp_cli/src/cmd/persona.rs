use jp_conversation::PersonaId;

use super::Output;
use crate::{error::Error, Ctx};

#[derive(Debug, clap::Args)]
pub struct Args {
    /// Persona name to switch to.
    pub name: String,
}

impl Args {
    pub fn run(self, ctx: &mut Ctx) -> Output {
        // Validate the persona exists
        let persona_id = PersonaId::try_from(&self.name)?;
        ctx.workspace
            .get_persona(&persona_id)
            .ok_or(Error::NotFound("Persona", self.name.clone()))?;

        // Update context with new persona
        ctx.workspace
            .get_active_conversation_mut()
            .context
            .persona_id = persona_id;

        Ok(format!("Switched to persona: {}", self.name).into())
    }
}
