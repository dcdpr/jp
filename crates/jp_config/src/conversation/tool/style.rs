//! Display style configuration for tools.

use std::{fmt, num::ParseIntError};

use schematic::{Config, ConfigEnum};
use serde::{Deserialize, Serialize};

use crate::assignment::{missing_key, AssignKeyValue, AssignResult, KvAssignment};

/// Display style configuration.
#[derive(Debug, Clone, PartialEq, Config)]
#[config(rename_all = "snake_case")]
pub struct DisplayStyleConfig {
    /// How to display the results of the tool call.
    #[setting(default)]
    pub inline_results: InlineResults,

    /// How to display the link to the file containing the tool call results.
    #[setting(default)]
    pub results_file_link: LinkStyle,
}

impl AssignKeyValue for PartialDisplayStyleConfig {
    fn assign(&mut self, kv: KvAssignment) -> AssignResult {
        match kv.key_string().as_str() {
            "" => *self = kv.try_object()?,
            "inline_results" => self.inline_results = kv.try_some_from_str()?,
            "results_file_link" => self.results_file_link = kv.try_some_from_str()?,
            _ => return missing_key(&kv),
        }

        Ok(())
    }
}

/// Whether and how to show the tool call results inline in the terminal.
///
/// Even if disabled or truncated, a link will be added to a file containing the
/// full tool call results. Additionally, the full tool call results will be
/// sent back to the assistant, regardless of this setting.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ConfigEnum)]
#[serde(rename_all = "snake_case")]
pub enum InlineResults {
    /// Never show the tool call results inline.
    Off,

    /// Show the full tool call results inline.
    Full,

    /// Show the first N lines of the tool call results inline.
    #[variant(fallback)]
    Truncate(Truncate),
}

impl Default for InlineResults {
    fn default() -> Self {
        Self::Truncate(Truncate::default())
    }
}

/// Truncate the tool call results to the first N lines.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Truncate {
    /// The number of lines to show.
    pub lines: usize,
}

impl Default for Truncate {
    fn default() -> Self {
        Self { lines: 10 }
    }
}

impl TryFrom<&str> for Truncate {
    type Error = ParseIntError;

    fn try_from(s: &str) -> Result<Self, Self::Error> {
        s.parse().map(|lines| Self { lines })
    }
}

impl fmt::Display for Truncate {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.lines)
    }
}

/// How to display the link to the file containing the tool call results.
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize, ConfigEnum)]
#[serde(rename_all = "lowercase")]
pub enum LinkStyle {
    /// Full (raw) link.
    #[default]
    Full,

    /// Clickable link using the `osc8` escape sequence.
    Osc8,

    /// No link.
    Off,
}
