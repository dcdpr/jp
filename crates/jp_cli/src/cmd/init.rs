use std::{env, fs, path::PathBuf, str::FromStr as _};

use crossterm::style::Stylize as _;
use duct::cmd;
use jp_config::{
    conversation::tool::RunMode,
    model::id::{ModelIdConfig, Name, PartialModelIdConfig, ProviderId},
    PartialAppConfig,
};
use jp_workspace::Workspace;
use path_clean::PathClean as _;

use crate::{ctx::IntoPartialAppConfig, Output, DEFAULT_STORAGE_DIR};

#[derive(Debug, clap::Args)]
pub(crate) struct Init {
    /// Path to initialize the workspace at. Defaults to the current directory.
    path: Option<PathBuf>,
}

impl Init {
    pub(crate) fn run(&self) -> Output {
        let cwd = std::env::current_dir()?;
        let mut root = self
            .path
            .clone()
            .unwrap_or_else(|| PathBuf::from("."))
            .clean();

        if !root.is_absolute() {
            root = cwd.join(root);
        }

        fs::create_dir_all(&root)?;

        let storage = root.join(DEFAULT_STORAGE_DIR);
        let id = jp_workspace::Id::new();
        jp_id::global::set(id.to_string());

        let mut workspace =
            Workspace::new_with_id(root.clone(), id.clone()).persisted_at(&storage)?;

        id.store(&storage)?;

        workspace = workspace.with_local_storage()?;

        let mut config = default_config();
        if let Some(id) = default_model() {
            print!("Using model {}", id.to_string().bold().blue());
            let note = "  (to use a different model, update `.jp/config.toml`)".to_owned();
            println!("{}\n", note.grey().italic());

            config.assistant.model.id = PartialModelIdConfig {
                provider: Some(id.provider),
                name: Some(id.name),
            };
        }

        let data = toml::to_string_pretty(&config)?;
        fs::write(storage.join("config.toml"), data)?;
        fs::create_dir_all(storage.join("config.d"))?;

        workspace.persist()?;

        let loc = if root == cwd {
            "current directory".to_owned()
        } else {
            root.to_string_lossy().bold().to_string()
        };

        Ok(format!("Initialized workspace at {loc}").into())
    }
}

#[expect(clippy::too_many_lines)]
fn default_config() -> jp_config::PartialAppConfig {
    let mut cfg = jp_config::PartialAppConfig::default();
    cfg.extends
        .get_or_insert_default()
        .push("config.d/**/*".into());

    // This is a required field without a default value (that is, the
    // `ToolsDefaultsConfig` type does not set a default value for `run`).
    //
    // By setting it explicitly, we ensure that the default generated config
    // file has this value set, which exposes it to the user. This is desired,
    // as this is an important security feature, which we don't want users to
    // have to rely on a default value that might change in the future.
    cfg.conversation.tools.defaults.run = Some(RunMode::Ask);

    if has_anthropic() {
        cfg.providers.llm.aliases.extend([
            ("anthropic".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Anthropic),
                name: Some(Name("claude-sonnet-4-0".into())),
            }),
            ("claude".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Anthropic),
                name: Some(Name("claude-sonnet-4-0".into())),
            }),
            ("sonnet".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Anthropic),
                name: Some(Name("claude-sonnet-4-0".into())),
            }),
            ("opus".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Anthropic),
                name: Some(Name("claude-opus-4-1".into())),
            }),
            ("haiku".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Anthropic),
                name: Some(Name("claude-3-5-haiku-latest".into())),
            }),
        ]);
    }

    if has_openai() {
        cfg.providers.llm.aliases.extend([
            ("openai".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5".into())),
            }),
            ("chatgpt".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5".into())),
            }),
            ("gpt".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5".into())),
            }),
            ("gpt5".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5".into())),
            }),
            ("gpt5-mini".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5-mini".into())),
            }),
            ("gpt-mini".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5-mini".into())),
            }),
            ("gpt5-nano".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5-nano".into())),
            }),
            ("gpt-nano".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5-nano".into())),
            }),
            ("o3-research".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("o3-deep-research".into())),
            }),
            ("o4-mini-research".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("o4-mini-deep-research".into())),
            }),
            ("codex".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5-codex".into())),
            }),
            ("gpt-5-codex".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("gpt-5-codex".into())),
            }),
            ("codex-mini".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Openai),
                name: Some(Name("codex-mini-latest".into())),
            }),
        ]);
    }

    if has_google() {
        cfg.providers.llm.aliases.extend([
            ("google".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Google),
                name: Some(Name("gemini-2.5-pro".into())),
            }),
            ("gemini".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Google),
                name: Some(Name("gemini-2.5-pro".into())),
            }),
            ("gemini-pro".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Google),
                name: Some(Name("gemini-2.5-pro".into())),
            }),
            ("gemini-flash".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Google),
                name: Some(Name("gemini-2.5-flash".into())),
            }),
            ("gemini-lite".to_owned(), PartialModelIdConfig {
                provider: Some(ProviderId::Google),
                name: Some(Name("gemini-2.5-flash-lite".into())),
            }),
        ]);
    }

    cfg
}

fn has_anthropic() -> bool {
    env::var("ANTHROPIC_API_KEY").is_ok()
}

fn has_openai() -> bool {
    env::var("OPENAI_API_KEY").is_ok()
}

fn has_google() -> bool {
    env::var("GOOGLE_API_KEY").is_ok()
}

fn default_model() -> Option<ModelIdConfig> {
    env::var("JP_CFG_ASSISTANT_MODEL_ID")
        .ok()
        .and_then(|v| ModelIdConfig::from_str(&v).ok())
        .or_else(|| {
            let models = cmd!("ollama", "list")
                .pipe(cmd!("cut", "-d", " ", "-f1"))
                .pipe(cmd!("tail", "-n+2"))
                .read()
                .unwrap_or_default();

            let models = models.lines().map(str::trim).collect::<Vec<_>>();
            let model = if let Some(model) = models.iter().find(|m| m.starts_with("llama")) {
                model
            } else if let Some(model) = models.iter().find(|m| m.starts_with("gemma")) {
                model
            } else if let Some(model) = models.iter().find(|m| m.starts_with("qwen")) {
                model
            } else {
                return None;
            };

            format!("ollama/{model}").parse().ok()
        })
        // TODO: Use `Config` env vars here.
        .or_else(|| {
            env::var("ANTHROPIC_API_KEY")
                .is_ok()
                .then(|| "anthropic/claude-sonnet-4-0".parse().ok())
                .flatten()
        })
        .or_else(|| {
            env::var("OPENAI_API_KEY")
                .is_ok()
                .then(|| "openai/o4-mini".parse().ok())
                .flatten()
        })
        .or_else(|| {
            env::var("GEMINI_API_KEY")
                .is_ok()
                .then(|| "google/gemini-2.5-flash-preview-05-20".parse().ok())
                .flatten()
        })
}

impl IntoPartialAppConfig for Init {
    fn apply_cli_config(
        &self,
        _workspace: Option<&Workspace>,
        partial: PartialAppConfig,
    ) -> std::result::Result<PartialAppConfig, Box<dyn std::error::Error + Send + Sync>> {
        Ok(partial)
    }
}

#[cfg(test)]
mod tests {
    use serial_test::serial;

    use super::*;

    pub(crate) struct EnvVarGuard {
        name: String,
        original_value: Option<String>,
    }

    impl EnvVarGuard {
        pub fn set(name: &str, value: &str) -> Self {
            let name = name.to_string();
            let original_value = std::env::var(&name).ok();
            unsafe { std::env::set_var(&name, value) };
            Self {
                name,
                original_value,
            }
        }
    }

    impl Drop for EnvVarGuard {
        fn drop(&mut self) {
            if let Some(ref original) = self.original_value {
                unsafe { std::env::set_var(&self.name, original) };
            } else {
                unsafe { std::env::remove_var(&self.name) };
            }
        }
    }

    #[test]
    #[serial(env_vars)]
    fn test_default_config() {
        let _env1 = EnvVarGuard::set("ANTHROPIC_API_KEY", "foo");
        let _env2 = EnvVarGuard::set("OPENAI_API_KEY", "bar");
        let _env3 = EnvVarGuard::set("GOOGLE_API_KEY", "baz");

        let config = default_config();

        insta::assert_toml_snapshot!(config);
    }
}
