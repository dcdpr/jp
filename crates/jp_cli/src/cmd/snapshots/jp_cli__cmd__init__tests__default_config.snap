---
source: crates/jp_cli/src/cmd/init.rs
expression: config
---
extends = ['config.d/**/*']
[conversation.tools."*"]
run = 'ask'
[providers.llm.aliases.anthropic]
provider = 'anthropic'
name = 'claude-sonnet-4-0'

[providers.llm.aliases.claude]
provider = 'anthropic'
name = 'claude-sonnet-4-0'

[providers.llm.aliases.sonnet]
provider = 'anthropic'
name = 'claude-sonnet-4-0'

[providers.llm.aliases.opus]
provider = 'anthropic'
name = 'claude-opus-4-1'

[providers.llm.aliases.haiku]
provider = 'anthropic'
name = 'claude-3-5-haiku-latest'

[providers.llm.aliases.openai]
provider = 'openai'
name = 'gpt-5'

[providers.llm.aliases.chatgpt]
provider = 'openai'
name = 'gpt-5'

[providers.llm.aliases.gpt]
provider = 'openai'
name = 'gpt-5'

[providers.llm.aliases.gpt5]
provider = 'openai'
name = 'gpt-5'

[providers.llm.aliases.gpt5-mini]
provider = 'openai'
name = 'gpt-5-mini'

[providers.llm.aliases.gpt-mini]
provider = 'openai'
name = 'gpt-5-mini'

[providers.llm.aliases.gpt5-nano]
provider = 'openai'
name = 'gpt-5-nano'

[providers.llm.aliases.gpt-nano]
provider = 'openai'
name = 'gpt-5-nano'

[providers.llm.aliases.o3-research]
provider = 'openai'
name = 'o3-deep-research'

[providers.llm.aliases.o4-mini-research]
provider = 'openai'
name = 'o4-mini-deep-research'

[providers.llm.aliases.codex]
provider = 'openai'
name = 'gpt-5-codex'

[providers.llm.aliases.gpt-5-codex]
provider = 'openai'
name = 'gpt-5-codex'

[providers.llm.aliases.codex-mini]
provider = 'openai'
name = 'codex-mini-latest'

[providers.llm.aliases.google]
provider = 'google'
name = 'gemini-pro-latest'

[providers.llm.aliases.gemini]
provider = 'google'
name = 'gemini-pro-latest'

[providers.llm.aliases.gemini-pro]
provider = 'google'
name = 'gemini-pro-latest'

[providers.llm.aliases.gemini-flash]
provider = 'google'
name = 'gemini-flash-latest'

[providers.llm.aliases.gemini-lite]
provider = 'google'
name = 'gemini-flash-lite-latest'
